### PR TITLE
add auth token param converter

### LINKED ISSUE #477: Implement BearerTokenParamConverter that converts IAE -> ForbiddenException
moved from internal issue witchcraft#346

### PR DESCRIPTION
**Pull request by bavardage**
fixes #477 

This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/478)

### CODE REVIEW

File: jersey-servers/build.gradle [hunk 1]
```diff
@@ -9,6 +9,7 @@ dependencies {
     compile "com.fasterxml.jackson.module:jackson-module-afterburner"
     compile "com.jcraft:jzlib"
     compile "com.palantir.safe-logging:safe-logging"
+    compile "com.palantir.tokens:auth-tokens"
```
Comment (bavardage) [inline jersey-servers/build.gradle:12]: don't like this so much... not sure if there's a better place for param converter to live..?

File: jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/JsonExceptionMapper.java [hunk 2]
```diff
@@ -88,6 +90,14 @@ public final Response toResponse(T exception) {
             builder.type(MediaType.TEXT_PLAIN);
             builder.entity(exceptionMessage);
         }
+
+        if (exception instanceof NotAuthorizedException) {
```
Comment (bavardage) [inline jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/JsonExceptionMapper.java:94]: I could build a separate `NotAuthorizedExceptionMapper` but it wasn't obvious how to do that without repeating all of `#toResponse`...

Comment (bavardage) [reply jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/JsonExceptionMapper.java:94]: aside: of the subtypes of `ClientErrorException`, only `NotAuthorizedException` and `NotAllowedException` seem to require special handling (i.e. some response header). I'm not convinced we're probably tackling `NotAllowedException` (https://httpstatuses.com/405)

File: jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/AuthHeaderParamConverterProvider.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public static final class AuthHeaderParamConverter implements ParamConverter<AuthHeader> {
+        @Override
+        public AuthHeader fromString(final String value) {
+            try {
+                return AuthHeader.valueOf(value);
+            } catch (IllegalArgumentException e) {
+                // rethrow as forbidden with 'Bearer' challenge
+                throw new NotAuthorizedException(e, "Bearer");
```
Comment (uschi2000) [inline jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/AuthHeaderParamConverterProvider.java]: do we really need all this stuff? in my mind, sending a malformed auth header is just another IAE.
How would a (backend) client respond to a NAE+"Bearer" challenge?

**Comment by bavardage** [comment]
circle build fails because unable to find baseline...?

```
* What went wrong:
Execution failed for task ':resolveConfigurations'.
> Could not resolve all dependencies for configuration ':baseline'.
   > Could not find com.palantir:gradle-baseline-java-config:0.9.2.
     Searched in the following locations:
         https://jcenter.bintray.com/com/palantir/gradle-baseline-java-config/0.9.2/gradle-baseline-java-config-0.9.2.pom
         https://jcenter.bintray.com/com/palantir/gradle-baseline-java-config/0.9.2/gradle-baseline-java-config-0.9.2.zip
     Required by:
         project :

```

**Comment by bavardage** [comment]
https://jcenter.bintray.com/com/palantir/gradle-baseline-java-config/ baseline 0.9.2 doesn't seem to show up here?

**Comment by uschi2000** [comment]
Add the bintray bucket directly, see readme.

Sent from Email+ secured by MobileIron

-------- Original Message --------

**Comment by bavardage** [comment]
@uschi2000 made this build - should this be pointed at develop or at 3.0?

**Comment by uschi2000** [comment]
3.0, please

**Comment by bavardage** [comment]
WIP, blocked on https://github.com/palantir/http-remoting-api/issues/29

File: jersey-servers/build.gradle (hunk 1 above)
Comment (markelliot) [reply jersey-servers/build.gradle:12]: also don't we want auth-tokens2?

one option is to add the param converter to auth-tokens...

**Comment by bavardage** [comment]
ok moving this to auth tokens as a filter (not a param converter)

File: jersey-servers/build.gradle (hunk 1 above)
Comment (bavardage) [reply jersey-servers/build.gradle:12]: yeah, locally moved over but then hit the blocker that there's not an ErrorType for a 401.

Will at some point add code instead to auth tokens, not as param converter though. There's no (non deprecated) remoting support any more for throwing an exception that'll get mapped into the right thing, so instead I'll have to make a filter

**Comment by markelliot** [comment]
auth-tokens2, not http-remoting3. The former is a new version of the tokens you're intercepting.

On Aug 10, 2017, at 03:04, Ben Duffield <notifications@github.com<mailto:notifications@github.com>> wrote:


@bavardage commented on this pull request.

________________________________

In jersey-servers/build.gradle[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_http-2Dremoting_pull_478-23discussion-5Fr132412860&d=DwMCaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=NC5Ik0Xberkh-KLFSY8YV_ba92dfh619VLS3aqDtl58&s=MkClTK5PIlniQ9ZJYZCayH7DX_Js1UjJHEVwSXBLV4U&e=>:

> @@ -9,6 +9,7 @@ dependencies {
     compile "com.fasterxml.jackson.module:jackson-module-afterburner"
     compile "com.jcraft:jzlib"
     compile "com.palantir.safe-logging:safe-logging"
+    compile "com.palantir.tokens:auth-tokens"


yeah, locally moved over but then hit the blocker that there's not an ErrorType for a 401.

Will at some point add code instead to auth tokens, not as param converter though. There's no (non deprecated) remoting support any more for throwing an exception that'll get mapped into the right thing, so instead I'll have to make a filter

**Comment by bavardage** [comment]
Yes. Locally I had changes which did the move to auth tokens 2, and then I started on moving over from the removed exception mapper and realized the new exceptions don’t do unauthorized/provide any place to stick a challenge and so abandoned the approach/didn’t push.

 

I will look at instead adding a filter to auth-tokens2.

**Pull request closed without merging**